### DISCUSSION TITLE
Is it ok to `Module.callMain` with ASYNCIFY?

### DISCUSSION THREAD

**Discussion by iliakan** [emscripten-core/emscripten#18792] [Q&A] [answered]
I've read in the docs that one should use `ccall` with `-sASINCIFY` to call a C function that has nested async calls.

Does it mean that `Module.callMain` shouldn't be used with `ASYNCIFY`, and I should replace it with `ccall`?

P.S. I use `callMain` and I get Uncaught (in promise) `ExitStatus {name: 'ExitStatus', message: 'Program terminated with exit(0)', status: 0}`, originating from `_proc_exit`, _after_ `callMain` completes.

**Comment by iliakan** [comment]
So to say, `callMain` did work for a simple case, but gave that error.

I tried to replace `await Module.callMain(args)` with `await Module.ccall("main", "number", ["number", "number"], [argc, argv], {async: true})`.

It again kind of works, but there's the error is somewhere in the middle:
```
program exited (with status: 0), but keepRuntimeAlive() is set (counter=1) due to an async operation, 
so halting execution but not exiting the runtime or preventing further async execution 
(you can use emscripten_force_exit, if you want to force a true shutdown)
```

Does it mean I forgot to insert `await` somewhere? Any clue? ;)

**Reply by sbc100** [reply]
I don't think that log message is anything to worry about.. we should probably not display it by default.

**Reply by iliakan** [reply]
Did I get it right that `await Module.callMain` is a bad thing? 
There's some async magic in `ccall`, but really not sure how important it is.

**Comment by iliakan** [comment, marked as answer]
I removed `-O0`, and now every `await Module.ccall` throws an exception:

```
Uncaught (in promise) ExitStatus {name: 'ExitStatus', message: 'Program terminated with exit(0)', status: 0}
```

Am I supposed to wrap every `await Module.ccall` into try/catch and silence such exceptions?

**Reply by sbc100** [reply]
`callMain` is really an internal function for emscripten.   Normally it is run by `run` function.   `callMain` takes case of calling your main function and also shutting down the runtime after main returns.  The last thing is does is call `exitJS` which is turn calls `proc_exit` which will throw `ExitStatus`.

See `src/library.js` for the definitions of these various functions. 

So yes, anyone that calls `main` or `callMain` should be prepared to handle the `ExitStatus` exception.   Its now an emscripten program terminates itself and unwinds the stack.

**Reply by sbc100** [reply]
Actually calling in Module export all could result in `ExitStatus` being thrown, so one always needs to be prepared to handle that I guess.

**Reply by iliakan** [reply]
Thank you. The `ExitStatus` exception was getting out only with ASYNCIFY for me. 

Isn't it specifically made for async code?
